internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,574 @@
+	return nil
+}
+
+func (ec *escalationChecker) checkClusterRoleBindingEscalation(ctx context.Context, clusterRoleBinding *rbacv1.ClusterRoleBinding) error {
+	if rbacregistry.EscalationAllowed(ctx) {
+		return nil
+	}
+
+	roleRef := rbacinternal.RoleRef{}
+	err := rbacv1helpers.Convert_v1_RoleRef_To_rbac_RoleRef(&clusterRoleBinding.RoleRef, &roleRef, nil)
+	if err != nil {
+		return err
+	}
+
+	if rbacregistry.BindingAuthorized(ctx, roleRef, metav1.NamespaceNone, ec.authorizer) {
+		return nil
+	}
+
+	rules, err := ec.ruleResolver.GetRoleReferenceRules(ctx, clusterRoleBinding.RoleRef, metav1.NamespaceNone)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if clusterRoleBinding.RoleRef.Kind == "ClusterRole" {
+		if manifestClusterRole, ok := ec.extraClusterRoles[types.NamespacedName{Name: clusterRoleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ec *escalationChecker) checkRoleEscalation(ctx context.Context, role *rbacv1.Role) error {
+	if rbacregistry.EscalationAllowed(ctx) || rbacregistry.RoleEscalationAuthorized(ctx, ec.authorizer) {
+		return nil
+	}
+
+	rules := role.Rules
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ec *escalationChecker) checkRoleBindingEscalation(ctx context.Context, roleBinding *rbacv1.RoleBinding) error {
+	if rbacregistry.EscalationAllowed(ctx) {
+		return nil
+	}
+
+	roleRef := rbacinternal.RoleRef{}
+	err := rbacv1helpers.Convert_v1_RoleRef_To_rbac_RoleRef(&roleBinding.RoleRef, &roleRef, nil)
+	if err != nil {
+		return err
+	}
+	if rbacregistry.BindingAuthorized(ctx, roleRef, roleBinding.Namespace, ec.authorizer) {
+		return nil
+	}
+
+	rules, err := ec.ruleResolver.GetRoleReferenceRules(ctx, roleBinding.RoleRef, roleBinding.Namespace)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	switch roleRef.Kind {
+	case "ClusterRole":
+		if manifestClusterRole, ok := ec.extraClusterRoles[types.NamespacedName{Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	case "Role":
+		if manifestRole, ok := ec.extraRoles[types.NamespacedName{Namespace: roleBinding.Namespace, Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+var fullAuthority = []rbacv1.PolicyRule{
+	{Verbs: []string{"*"}, APIGroups: []string{"*"}, Resources: []string{"*"}},
+	{Verbs: []string{"*"}, NonResourceURLs: []string{"*"}},
+}
+
+func parseEscalationErrorForMissingRules(ecError error) ([]rbacv1.PolicyRule, error) {
+	errRegex := regexp.MustCompile(`(?s)^(user \".*\" \(groups=.*\) is attempting to grant RBAC permissions not currently held):.*?$`)
```

Comment:
@joelanford is this something like what you were imagining for the regex?

Reply:
It seemed this is where @joelanford was looking for a new, type-assertable error structure?

Reply:
I imagine that where https://github.com/kubernetes/kubernetes/pull/130955 will play a role, if we can get it merged (and when we get to that version of k8s).

Reply:
This worries me a little. Do we always expect matches? What happens if there are no matches? Will we NPE `errMatches[0][1]` (line 560)?
Should we also add a comment explaining the returned error is part of the parsing output (rather than an error in the parsing itself)?

Reply:
done in: f9721fbf